### ISSUE TITLE
Downloads fail if any parent directories are missing

### ISSUE THREAD

**Issue by crwood** [LeastAuthority/magic-folder#571]
When testing new-magic-folder recovery for Gridsync, I noticed that, sometimes, when I attempt to re-join/restore a magic-folder that was joined previously, the expected files never actually re-download (leaving Magic-Folder to essentially do nothing indefinitely).

I last observed this phenomena with a test directory structure that looked something like this locally:

```
Cat Pics/lolcat.jpg
Cat Pics/Nala.jpg
Cat Pics/subdir/Garfield.jpg
```

After confirming both a) that the remote files indeed existed on grid (i.e., that they were correctly uploaded previously) and b) that Magic-Folder had indeed re-joined the folder in question (i.e., that the post-restoragy configuration was pointing to the correct caps and the new participant was successfully added to the collective), I noticed these websocket messages in my logs:

```
{"state": {"folders": {"Cat Pics": {"downloads": [{"name": "subdir/Garfield.jpg", "started-at": 1634411390.386539}], "errors": [{"timestamp": 1634411390, "summary": "Failed to overwrite file 'subdir/Garfield.jpg': [Errno 2] No such file or directory"}], "uploads": [], "recent": []}}, "synchronizing": true}}
{"state": {"folders": {"Cat Pics": {"downloads": [], "errors": [{"timestamp": 1634411390, "summary": "Failed to overwrite file 'subdir/Garfield.jpg': [Errno 2] No such file or directory"}], "uploads": [], "recent": []}}, "synchronizing": false}}
```

(For what it's worth, if I start from a completely fresh configuration and only add directories to Magic-Folder that do not contain subdirectories, downloads for re-joined/restored folders work as expected.)

From what I can tell, Magic-Folder thus 1) will fail to download a given remote file if the parent directories for that filepath do not already exist locally and 2) will not download any/all other files for a given magic-folder if there is one such failure (in my case, `lolcat.jpg` and `Nala.jpg` never downloaded, despite being in the "top-level" of the magic-folder).

Regarding issue "1)" above, I would expect that Magic-Folder should attempt to create (or otherwise ensure the existence of) the local directory/path for a given file -- including any missing parent directories -- before attempting to download a file to that path (rather than throwing an error).

Regarding problem "2)", I would expect that Magic-Folder would continue to attempt to download any other pending files, even if one of the other downloads have failed.

**Comment by meejah** [comment]
Yeah, the ultimate problem here I think is "just" that paths leading to a newly-noticed remote file do not get created (locally). I have a fix for this on a branch with some other changes though.

One "gotcha" is what to do if there's already a local file where a directory "should" be .. or if a file `foo` exists, then is deleted, then becomes a directory later. This issues are for remote-sync. cases though.
